enhancements/authentication/allowing-uri-scheme-in-oidc-sub-claims.md

```diff
@@ -0,0 +1,167 @@
+`Identity` object that describes the mapping between the identity provider the
+user is a member of, and the user.
+
+The name of the `Identity` object is constructed by using the following pattern:
+
+```text
+<identity provider name from config>:<unique user identifier from the identity provider>
+```
+
+This naming is further validated so that neither the username, nor the identity provider
+name may contain the `:` character, and, according to Kubernetes naming schemes,
+neither can also contain the `/` sign, yet both of these commonly appear as a
+part of URI/URL schemes.
+
+## Proposal
+
+Due to the API limitations mentioned in [Problem Summary](#problem-summary), a
+generic `sub` claim cannot be safely used in order to create an `Identity` object.
+It is apparent that these will need to be encoded in a certain way. Simply encoding
+a `sub` claim won't do as that would break authentication for any OIDC user
+that had its `Identity` created prior to the version containing this fix.
+
+Proposal: if a `sub` claim of an ID token contains either `:` or `/`, url-encode
+the value of the claim and create the new identity object like so:
+
+```yaml
+apiVersion: user.openshift.io/v1
+kind: Identity
+metadata:
+  name: <identity provider name>:url:<url-encoded sub claim value>
+providerName: <identity provider name>
+providerUserName: <url-encoded sub claim value>
+user:
+  name: <preferred username || url-encoded sub claim value>
+  uid: <as usual>
+```
+
+Notice the naming scheme change where the `Identity` object name now consists of
+three sections instead of previous two. The additional section simply contains
+`url` to mark the encoding scheme used in the third section of the name.
+
+The username of the user object also needs to be an encoded `sub` claim value if
+a preferred username is not configured for the identity provider.
+
+### User Stories
+
+#### Story 1
+
+I am a user of an OpenID Connect identity provider and I would like to log in to
+OpenShift.
+
+### Implementation Details
+
+#### oauth-server
+
+When creating a user and an identity object, the oauth-server will have to check
+whether the username it retrieved from the third-party authentication logic contains
+either of `:` or `/`, and if it does, follow the naming scheme from [Proposal](#proposal),
+that is, url-encode the username for `User` objects for the object name, and
+url-encode the username and prepend it with `url:` to use as a username in the
+`Identity` object naming scheme.
+
+#### oauth-apiserver
+
+Validation of the names of the `Identity` objects need to be loosened in order
+to allow three sections separated by the `:` character, where the only allowed
+value for the second section is `url`.
+
+### Risks and Mitigations
+
+When a cluster with existing `Identity` objects of the new naming schemes is downgraded,
+logging in with OIDC providers using URI scheme is their ID tokens' `sub` claims
+will be broken again.
```

Comment:
tldr: this seems to be acceptable.

generally, afaik the idiom is that downgrades should generally not draw the cluster dysfunctional as downgrade scenarios are mostly used as temporary timeframes to upgrade again so this seems acceptable.

but to be sure: is it possible that those oidc id tokens appear in some form of service-to-service communication within our control plane or user workloads? I don't think so but just asking to make sure.

Reply:
There is no way to keep the authentication working for the afflicted clusters on downgrade as they would not work prior to the upgrade either. An upgrade means a switch from "this has always been broken" to fixed.

The ID tokens should only be visible to the oauth-server and none other.